From cf8b18ecab269653424df207c121eaaaf9f7808b Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 01 Dec 2021 17:38:43 +0200
Subject: [PATCH] [SW-36216] gaudi2: sec: add dcore rtr for pcie

In gaudi2, apart from the pcie mstr if, pcie traffic is also routed
through dcore0 rtr0 if. Hence configuring pcie mstr if range
registers alone is not enough to block/allow pcie traffic.

This patch couples the dcore0 rtr0 if along with pcie rr
configuration such that the common is reused.

Change-Id: Iba559d3e9b96f52da14e2eb0e7597425cf0783c1
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index 39ec6f3..ca953e1 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -37,7 +37,6 @@
  * hit_wr_breg : base register for hit read.
  */
 struct hl_sec_rrs_cfg {
-	int instances_num;
 	u32 instances_dist;
 	u32 min_short_breg;
 	u32 max_short_breg;
@@ -70,7 +69,6 @@
 };
 
 struct hl_sec_rrs_cfg_hbw {
-	int instances_num;
 	u32 instances_dist;
 	struct hbw_rr_addr min_short;
 	struct hbw_rr_addr max_short;
@@ -189,10 +187,14 @@
 #define SFT_INST_NUM	1
 #define SHIFT_4K_POS	12
 #define PCIE_HBW_NUM	1
+/*
+ * Number of RTRs which allow pcie traffic
+ * (DCORE0 RTR0 for gaudi2)
+ */
+#define PCIE_RTR_NUM	1
 
 static struct hl_sec_rrs_cfg hl_sec_lbw_rrs_tbl[] = {
 	[HL_SEC_ENGINE_IF_RTR] = {
-		.instances_num = NUM_OF_RTR_PER_DCORE,
 		.instances_dist = IF_RTR_DIST,
 		.min_short_breg = mmDCORE0_RTR0_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
 		.max_short_breg = mmDCORE0_RTR0_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
@@ -202,7 +204,6 @@
 		.hit_rd_breg = mmDCORE0_RTR0_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
 	},
 	[HL_SEC_ENGINE_PCIE] = {
-		.instances_num = PCIE_INST_NUM,
 		.instances_dist = PCIE_INST_DIST,
 		.min_short_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
 		.max_short_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
@@ -212,7 +213,6 @@
 		.hit_rd_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
 	},
 	[HL_SEC_ENGINE_PSOC] = {
-		.instances_num = PSOC_INST_NUM,
 		.instances_dist = 0x1000,
 		.min_short_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
 		.max_short_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
@@ -222,7 +222,6 @@
 		.hit_rd_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
 	},
 	[HL_SEC_ENGINE_SFT] = {
-		.instances_num = SFT_INST_NUM,
 		.instances_dist = 0,
 		.min_short_breg = mmSFT0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
 		.max_short_breg = mmSFT0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
@@ -238,7 +237,6 @@
 
 static struct hl_sec_rrs_cfg_hbw hl_sec_hbw_rrs_tbl[] = {
 	[HL_SEC_ENGINE_IF_RTR] = {
-		.instances_num = NUM_OF_RTR_PER_DCORE,
 		.instances_dist = IF_RTR_DIST,
 		.min_short.lo = mmDCORE0_RTR0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_LO_0,
 		.min_short.hi = mmDCORE0_RTR0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_HI_0,
@@ -252,7 +250,6 @@
 		.hit_rd_breg = mmDCORE0_RTR0_MSTR_IF_RR_SHRD_HBW_PRIV_HIT_AR,
 	},
 	[HL_SEC_ENGINE_PCIE] = {
-		.instances_num = PCIE_HBW_NUM,
 		.instances_dist = 0,
 		.min_short.lo = mmPCIE_MSTR_RR_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_LO_0,
 		.min_short.hi = mmPCIE_MSTR_RR_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_HI_0,
@@ -266,7 +263,6 @@
 		.hit_rd_breg = mmPCIE_MSTR_RR_MSTR_IF_RR_SHRD_HBW_PRIV_HIT_AR,
 	},
 	[HL_SEC_ENGINE_PSOC] = {
-		.instances_num = PSOC_INST_NUM,
 		.instances_dist = PSOC_INST_DIST,
 		.min_short.lo = mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_LO_0,
 		.min_short.hi = mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_HI_0,
@@ -280,7 +276,6 @@
 		.hit_rd_breg = mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_HIT_AR,
 	},
 	[HL_SEC_ENGINE_SFT] = {
-		.instances_num = HBW_SFT_INST_NUM,
 		.instances_dist = HBW_SFT_DIST,
 		.min_short.lo = mmSFT0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_LO_0,
 		.min_short.hi = mmSFT0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_HI_0,
@@ -349,7 +344,7 @@
 #define MAX_LBW_SHORT_RR	14
 
 static void hl_secure_lbw_rr_gen(int rr_idx, u64 min, u64 max, int dist,
-		enum hl_sec_engine_type type)
+		int inst_num, enum hl_sec_engine_type type)
 {
 	int i;
 	struct hl_sec_rrs_cfg *cfg = &hl_sec_lbw_rrs_tbl[type];
@@ -375,7 +370,7 @@
 	rr_reg.hit.hit_wr = cfg->hit_wr_breg + dist;
 	rr_reg.hit.hit_rd = cfg->hit_rd_breg + dist;
 
-	for (i = 0; i < cfg->instances_num; i++)
+	for (i = 0; i < inst_num; i++)
 		hl_secure_update_rr_reg(rr_idx, cfg->instances_dist * i, &rr_reg);
 }
 
@@ -386,18 +381,22 @@
 	if (rr_sel.noc) {
 		for (i = 0; i < NUM_OF_DCORES; i++) {
 			hl_secure_lbw_rr_gen(rr_idx, min, max, DCORE_DIST * i,
-					HL_SEC_ENGINE_IF_RTR);
+					NUM_OF_RTR_PER_DCORE, HL_SEC_ENGINE_IF_RTR);
 
 			hl_secure_lbw_rr_gen(rr_idx, min, max, SFT_DIST * i,
-					HL_SEC_ENGINE_SFT);
+					SFT_INST_NUM, HL_SEC_ENGINE_SFT);
 		}
 	}
 
-	if (rr_sel.pcie)
-		hl_secure_lbw_rr_gen(rr_idx, min, max, 0, HL_SEC_ENGINE_PCIE);
+	if (rr_sel.pcie) {
+		hl_secure_lbw_rr_gen(rr_idx, min, max, 0, PCIE_INST_NUM, HL_SEC_ENGINE_PCIE);
+		if (!rr_sel.noc)
+			hl_secure_lbw_rr_gen(rr_idx, min, max, 0, PCIE_RTR_NUM,
+						HL_SEC_ENGINE_IF_RTR);
+	}
 
 	if (rr_sel.psoc)
-		hl_secure_lbw_rr_gen(rr_idx, min, max, 0, HL_SEC_ENGINE_PSOC);
+		hl_secure_lbw_rr_gen(rr_idx, min, max, 0, PSOC_INST_NUM, HL_SEC_ENGINE_PSOC);
 }
 
 #define EN_ARC_SEC_PRIV				0x3
@@ -591,7 +590,7 @@
 }
 
 static void hl_secure_hbw_rr_gen(int rr_idx, u64 min, u64 max,
-		int dist, enum hl_sec_engine_type type)
+		int dist, int inst_num, enum hl_sec_engine_type type)
 {
 	int i;
 	struct hbw_rr_reg rr_reg;
@@ -617,7 +616,7 @@
 	rr_reg.hit.hit_wr = cfg->hit_wr_breg + dist;
 	rr_reg.hit.hit_rd = cfg->hit_rd_breg + dist;
 
-	for (i = 0; i < cfg->instances_num; i++)
+	for (i = 0; i < inst_num; i++)
 		hl_secure_update_rr_reg_hbw(rr_idx, cfg->instances_dist * i, &rr_reg);
 }
 
@@ -628,18 +627,31 @@
 	if (rr_sel.noc) {
 		for (i = 0; i < NUM_OF_DCORES; i++) {
 			hl_secure_hbw_rr_gen(rr_idx, min, max, DCORE_DIST * i,
-					HL_SEC_ENGINE_IF_RTR);
+					NUM_OF_RTR_PER_DCORE, HL_SEC_ENGINE_IF_RTR);
 
 			hl_secure_hbw_rr_gen(rr_idx, min, max, SFT_DIST * i,
-					HL_SEC_ENGINE_SFT);
+					HBW_SFT_INST_NUM, HL_SEC_ENGINE_SFT);
 		}
 	}
 
-	if (rr_sel.pcie)
-		hl_secure_hbw_rr_gen(rr_idx, min, max, 0, HL_SEC_ENGINE_PCIE);
+	if (rr_sel.pcie) {
+		hl_secure_hbw_rr_gen(rr_idx, min, max, 0, PCIE_HBW_NUM, HL_SEC_ENGINE_PCIE);
+		/* Apart from the pcie mstr if, pcie traffic is also routed
+		 * through dcore0 rtr0 if. Hence configuring pcie mstr if range
+		 * registers alone is not enough to block/allow pcie traffic.
+		 * The dcore0 rtr0 is therefore coupled with pcie rr configuration.
+		 */
+		/*
+		 * In case noc is already enabled along with pcie, then dcore0 rtr0 is configured
+		 * as a part of noc select itself, hence not required to reconfigure in such case.
+		 */
+		if (!rr_sel.noc)
+			hl_secure_hbw_rr_gen(rr_idx, min, max, 0, PCIE_RTR_NUM,
+						HL_SEC_ENGINE_IF_RTR);
+	}
 
 	if (rr_sel.psoc)
-		hl_secure_hbw_rr_gen(rr_idx, min, max, 0, HL_SEC_ENGINE_PSOC);
+		hl_secure_hbw_rr_gen(rr_idx, min, max, 0, PSOC_INST_NUM, HL_SEC_ENGINE_PSOC);
 }
 
 #define SPI_FLASH_SIZE	0x1000000ull /* 16MB */
@@ -726,12 +738,6 @@
 	memset(&rr_sel, 0, sizeof(struct hl_rr_eng_select));
 	/* Block pcie RR for dyn rr */
 	rr_sel.pcie = true;
-	/*
-	 * PCIE shares noc rtrs.
-	 * TODO: Refine based on specific
-	 * rtrs.
-	 */
-	rr_sel.noc = true;
 
 	hl_secure_hbw_rr(rr_idx, min64, max64, rr_sel);
 }
